pageserver/src/layered_repository.rs

```diff
@@ -726,38 +747,42 @@ impl Timeline for LayeredTimeline {
             let layer = self.get_layer_for_write(seg, lsn)?;
             layer.put_truncation(lsn, relsize % RELISH_SEG_SIZE)?;
         }
-
+        self.decrease_current_logical_size((relsize - oldsize) * BLCKSZ as u32);
         Ok(())
     }
 
     fn put_unlink(&self, rel: RelishTag, lsn: Lsn) -> Result<()> {
         trace!("put_unlink: {} at {}", rel, lsn);
 
         if rel.is_blocky() {
-            let oldsize_opt = self.get_relish_size(rel, self.last_valid_lsn.load())?;
-            if let Some(oldsize) = oldsize_opt {
-                let old_last_seg = if oldsize == 0 {
-                    0
-                } else {
-                    (oldsize - 1) / RELISH_SEG_SIZE
-                };
-
-                // Unlink all segments
-                for remove_segno in 0..=old_last_seg {
-                    let seg = SegmentTag {
-                        rel,
-                        segno: remove_segno,
+            match self.get_relish_size(rel, self.last_valid_lsn.load())? {
+                Some(oldsize) => {
+                    let old_last_seg = if oldsize == 0 {
+                        0
+                    } else {
+                        (oldsize - 1) / RELISH_SEG_SIZE
                     };
-                    let layer = self.get_layer_for_write(seg, lsn)?;
-                    layer.put_unlink(lsn)?;
+
+                    // Unlink all segments
+                    for remove_segno in 0..=old_last_seg {
+                        let seg = SegmentTag {
+                            rel,
+                            segno: remove_segno,
+                        };
+                        let layer = self.get_layer_for_write(seg, lsn)?;
+                        layer.put_unlink(lsn)?;
+                    }
+                    self.decrease_current_logical_size(oldsize * BLCKSZ as u32);
+                }
+                None => {
+                    warn!(
+                        "put_unlink called on non-existent relish {} at {}",
+                        rel, lsn
+                    );
                 }
-            } else {
-                warn!(
-                    "put_unlink called on non-existent relish {} at {}",
-                    rel, lsn
-                );
             }
         } else {
+            // TODO handle TwoPhase relishes
```

Comment:
Can you elaborate? Does this not work for TwoPhase relishes?

Reply:
Correct, currently when we unlink TwoPhase record we do not know it's size (and I'm not sure is it ok to just get it via get_relish_size). I'was thinking about collecting this info during gc runs. When I've talked to @kelvich about that and TwoPhase records in general he said that there should be special handling in gc for TwoPhase records (they have to live longer than expected). I saw PR's into object repository's/rocksdb repo gc handling about that, but I do not see anything special to TwoPhase records in layered repo gc handling, so decided to put a todo here. Probably this needs a separate issue

Reply:
I've revisited this and I think this issue is not applicable to layered repo, so this is just because there is no TwoPhase size anywhere nearby in the unlinking stage